File: backend/F1Fantasy/Models/Result.cs
namespace F1Fantasy.Models;

public class Result
{
    public int Id { get; set; } // Auto-increment primary key
    public string Season { get; set; } = string.Empty;
    public string Round { get; set; } = string.Empty;
    public string Number { get; set; } = string.Empty;
    public string Position { get; set; } = string.Empty;
    public string PositionText { get; set; } = string.Empty;
    public string Points { get; set; } = string.Empty;
    public string DriverId { get; set; } = string.Empty;
    public Driver Driver { get; set; } = new();
    public string ConstructorId { get; set; } = string.Empty;
    public Constructor Constructor { get; set; } = new();
    public string Grid { get; set; } = string.Empty;
    public string Laps { get; set; } = string.Empty;
    public string Status { get; set; } = string.Empty; // Status text (e.g., "Finished", "Engine")
    public string? StatusId { get; set; } // Foreign key to Status table (e.g., "1", "5")
    public ResultTime? Time { get; set; }
    public FastestLap? FastestLap { get; set; }
    public bool IsSprint { get; set; } = false; // Indicates if this is a sprint race result
}

public class ResultTime
{
    public string Millis { get; set; } = string.Empty;
    public string Time { get; set; } = string.Empty;
}

public class FastestLap
{
    public string Rank { get; set; } = string.Empty;
    public string Lap { get; set; } = string.Empty;
    public LapTime Time { get; set; } = new();
    public AverageSpeed? AverageSpeed { get; set; }
}

public class LapTime
{
    public string Time { get; set; } = string.Empty;
}

public class AverageSpeed
{
    public string Units { get; set; } = string.Empty;
    public string Speed { get; set; } = string.Empty;
}

// API Response models for Results endpoint
public class ResultApiResponse
{
    public ResultMRData? MRData { get; set; }
}

public class ResultMRData
{
    public string? Xmlns { get; set; }
    public string? Series { get; set; }
    public string? Url { get; set; }
    public string? Limit { get; set; }
    public string? Offset { get; set; }
    public string? Total { get; set; }
    public ResultRaceTable? RaceTable { get; set; }
}

public class ResultRaceTable
{
    public string? Season { get; set; }
    public List<RaceWithResults>? Races { get; set; }
}

public class RaceWithResults
{
    public string Season { get; set; } = string.Empty;
    public string Round { get; set; } = string.Empty;
    public string Url { get; set; } = string.Empty;
    public string RaceName { get; set; } = string.Empty;
    public Circuit Circuit { get; set; } = new();
    public string Date { get; set; } = string.Empty;
    public string Time { get; set; } = string.Empty;
    public List<Result>? Results { get; set; }
    public List<Result>? SprintResults { get; set; } // Sprint race results
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/entity/Result.java
package no.f1fantasy.entity;

import jakarta.persistence.*;
import lombok.Data;

@Entity
@Table(name = "Results",
        indexes = {
                @Index(name = "IX_Results_Season_Round", columnList = "Season, Round"),
                @Index(name = "IX_Results_DriverId", columnList = "DriverId"),
                @Index(name = "IX_Results_ConstructorId", columnList = "ConstructorId"),
                @Index(name = "IX_Results_StatusId", columnList = "StatusId")
        })
@Data
public class Result {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "Id")
    private Integer id;

    @Column(name = "Season", length = 10, nullable = false)
    private String season;

    @Column(name = "Round", length = 10, nullable = false)
    private String round;

    @Column(name = "Number", length = 10)
    private String number;

    @Column(name = "Position", length = 10)
    private String position;

    @Column(name = "PositionText", length = 10)
    private String positionText;

    @Column(name = "Points", length = 10)
    private String points;

    @Column(name = "DriverId", length = 100, nullable = false)
    private String driverId;

    @Column(name = "ConstructorId", length = 100, nullable = false)
    private String constructorId;

    /** Nullable — sprint races or races with incomplete data may have null grid position. */
    @Column(name = "Grid", length = 10)
    private String grid;

    @Column(name = "Laps", length = 10)
    private String laps;

    @Column(name = "Status", length = 100)
    private String status;

    @Column(name = "StatusId", length = 10)
    private String statusId;

    @Column(name = "IsSprint", nullable = false)
    private boolean isSprint;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "millis", column = @Column(name = "Time_Millis", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "Time_Time", length = 50))
    })
    private ResultTime resultTime;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "rank", column = @Column(name = "FastestLap_Rank", length = 10)),
            @AttributeOverride(name = "lap", column = @Column(name = "FastestLap_Lap", length = 10)),
            // Nested via FastestLap embeddable's own @AttributeOverrides
            @AttributeOverride(name = "lapTime.time", column = @Column(name = "FastestLap_Time_Time", length = 50)),
            @AttributeOverride(name = "averageSpeed.units", column = @Column(name = "FastestLap_AverageSpeed_Units", length = 10)),
            @AttributeOverride(name = "averageSpeed.speed", column = @Column(name = "FastestLap_AverageSpeed_Speed", length = 20))
    })
    private FastestLap fastestLap;
}
